lib/rate_map.c

```diff
@@ -0,0 +1,200 @@
+**
+** This file is part of msprime.
+**
+** msprime is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** msprime is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+#include "util.h"
+#include "msprime.h"
+
+void
+rate_map_print_state(rate_map_t *self, FILE *out)
+{
+    size_t j;
+
+    fprintf(out, "rate_map (%p):: size = %d\n", (void *) self, (int) self->size);
+    fprintf(out, "\tsequence_length = %.14g\n", rate_map_get_sequence_length(self));
+    fprintf(out, "\tindex\tposition\tvalue\n");
+    for (j = 0; j <= self->size; j++) {
+        fprintf(out, "\t%d\t%.14g\t%.14g\n", (int) j, self->position[j], self->rate[j]);
+    }
+}
+
+int MSP_WARN_UNUSED
+rate_map_alloc(rate_map_t *self, size_t size, double *position, double *rate)
+{
+    int ret = 0;
+    size_t j;
+    double sum;
+
+    memset(self, 0, sizeof(rate_map_t));
+    if (size < 1) {
+        ret = MSP_ERR_INSUFFICIENT_INTERVALS;
+        goto out;
+    }
+    /* Check the framing position */
+    if (position[0] != 0.0) {
+        ret = MSP_ERR_INTERVAL_MAP_START_NON_ZERO;
+        goto out;
+    }
+    self->rate = malloc((size + 1) * sizeof(*self->rate));
+    self->position = malloc((size + 1) * sizeof(*self->position));
+    self->cumulative_mass = malloc((size + 1) * sizeof(*self->cumulative_mass));
+    if (self->position == NULL || self->rate == NULL || self->cumulative_mass == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    self->size = size;
+    sum = 0;
+    for (j = 0; j <= size; j++) {
+        if (!isfinite(position[j])) {
+            ret = MSP_ERR_NONFINITE_INTERVAL_POSITION;
+            goto out;
+        }
+        self->cumulative_mass[j] = sum;
+        self->position[j] = position[j];
+        if (j < size) {
+            /* This condition and the requirement that position 0 = 0 means that
+             * we can't have negative values. */
+            if (position[j] >= position[j + 1]) {
+                ret = MSP_ERR_INTERVAL_POSITIONS_UNSORTED;
+                goto out;
+            }
+            if ((!isfinite(rate[j])) || rate[j] < 0) {
+                ret = MSP_ERR_BAD_RATE_VALUE;
+                goto out;
+            }
+            self->rate[j] = rate[j];
+            sum += (position[j + 1] - position[j]) * rate[j];
+        }
+    }
+    /* We should never touch this, but we initialise it for neatness. */
```

Comment:
Is "neatness" so that an off-by-one error when indexing `rate` will give us an easier-to-debug error than if it was just off the end of the array?

Reply:
Yeah,  it's mainly so we don't have to put in a special case for the ``print_state`` loop. Arguably we'd be better off leaving it uninitialised so that valgrind would catch any accesses off the end of the array, though.

Reply:
Happy to change it.

Reply:
Well, if it's easier to find bugs if it's shorter, and the issue is only in `print_state`, then I vote to make it shorter. If `rate` isn't shorter by one then you have to remember where the spurious value is - is it the first or the last element?

Reply:
OK, will do.

Reply:
Done in ce85e75